syntax = "proto3";

package opinionanalyzer;

// OPINION ANALYZER SERVICE
// Provides ML-based opinion analysis via gRPC

service OpinionAnalyzer {
    // MATCH OPINION TO TOPIC USING SENTENCE SIMILARITY
    rpc MatchOpinionToTopic (MatchRequest) returns (MatchResponse);

    // CLASSIFY A SINGLE OPINION TYPE
    rpc ClassifyOpinion (ClassifyRequest) returns (ClassifyResponse);

    // CLASSIFY MULTIPLE OPINIONS IN BATCH
    rpc ClassifyOpinionsBatch (BatchClassifyRequest) returns (BatchClassifyResponse);

    // GENERATE CONCLUSION FROM TOPIC AND OPINIONS
    rpc GenerateConclusion (ConclusionRequest) returns (ConclusionResponse);

    // FULL ANALYSIS PIPELINE
    rpc AnalyzeOpinions (AnalyzeRequest) returns (AnalyzeResponse);

    // HEALTH CHECK
    rpc HealthCheck (HealthRequest) returns (HealthResponse);
}

// MATCH REQUEST - FIND TOPICS FOR AN OPINION
message MatchRequest {
    string opinion_text = 1;
    int32 top_k = 2;
    float threshold = 3;
}

// TOPIC MATCH RESULT
message TopicMatch {
    string topic_id = 1;
    string topic_text = 2;
    float similarity = 3;
}

// MATCH RESPONSE
message MatchResponse {
    repeated TopicMatch matches = 1;
    bool success = 2;
    string error_message = 3;
}

// CLASSIFY REQUEST - SINGLE OPINION
message ClassifyRequest {
    string opinion_text = 1;
}

// CLASSIFY RESPONSE
message ClassifyResponse {
    string predicted_type = 1;
    int32 predicted_id = 2;
    map<string, float> probabilities = 3;
    bool success = 4;
    string error_message = 5;
}

// BATCH CLASSIFY REQUEST
message BatchClassifyRequest {
    repeated string opinion_texts = 1;
}

// BATCH CLASSIFY RESPONSE
message BatchClassifyResponse {
    repeated ClassifyResponse results = 1;
    bool success = 2;
    string error_message = 3;
}

// OPINION MESSAGE FOR CONCLUSION GENERATION
message Opinion {
    string text = 1;
    string type = 2;
}

// CONCLUSION REQUEST
message ConclusionRequest {
    string topic_text = 1;
    repeated Opinion opinions = 2;
    float temperature = 3;
    int32 max_tokens = 4;
}

// CONCLUSION RESPONSE
message ConclusionResponse {
    string conclusion = 1;
    bool success = 2;
    string error_message = 3;
}

// FULL ANALYSIS REQUEST
message AnalyzeRequest {
    string topic_text = 1;
    repeated string opinion_texts = 2;
    bool generate_conclusion = 3;
}

// CLASSIFIED OPINION RESULT
message ClassifiedOpinion {
    string text = 1;
    string predicted_type = 2;
    float confidence = 3;
}

// FULL ANALYSIS RESPONSE
message AnalyzeResponse {
    string topic_text = 1;
    repeated ClassifiedOpinion classified_opinions = 2;
    string conclusion = 3;
    bool success = 4;
    string error_message = 5;
}

// HEALTH REQUEST
message HealthRequest {}

// HEALTH RESPONSE
message HealthResponse {
    bool healthy = 1;
    string status = 2;
    map<string, bool> services = 3;
}
